### PR TITLE
DM-14857: Make MatchPPMb default.

### PR DESCRIPTION
**Pull request by jdswinbank**

### CODE REVIEW

File: python/lsst/meas/astrom/matchPessimisticB.py [hunk 1]
```diff
@@ -95,7 +98,7 @@ class MatchPessimisticBConfig(pexConfig.Config):
     matcherIterations = pexConfig.RangeField(
         doc="Number of softening iterations in matcher.",
         dtype=int,
-        default=5,
+        default=3,
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/matchPessimisticB.py]: I'll add in the comment on rebase. 5 was a really aggressive and would allo for softenings of 32 times the initial guess which is way to much. For the record, I've never seen the matcher succeed if it needs that many iterations. I believe the old matcher only softens by about a factor of 2.

File: python/lsst/meas/astrom/matchPessimisticB.py [hunk 2]
```diff
@@ -420,22 +435,19 @@ def _doMatch(self, refCat, sourceCat, wcs, refFluxField, numUsableSources,
...
 
-        # Create our matcher object.
-        pyPPMb = PessimisticPatternMatcherB(ref_array[:, :3], self.log)
-
         match_found = False
         # Start the iteration over our tolerances.
-        for soften_pattern in range(self.config.matcherIterations):
-            for soften_dist in range(self.config.matcherIterations):
+        for soften_dist in range(self.config.matcherIterations):
+            for soften_pattern in range(self.config.matcherIterations):
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/matchPessimisticB.py]: Will do.

File: python/lsst/meas/astrom/matchPessimisticB.py [hunk 3]
```diff
@@ -41,27 +41,30 @@ def __init__(self, maxMatchDist=None, autoMaxMatchDist=None,
...
             Reference to the position in the magnitude sorted source array
             where a successful pattern match was found.
-        failedPatternList : list of ints
+        failedPatternList : `list` of `int`s
             List of ints specifying indicies in the magnitude sourced source
             array to skip. These are skipped are previous iterations that are
             likely false positives due to the code having to soften after a
             pattern is matched.
+        PPMbObj : `lsst.meas.astrom.PessimisticPatternMatcherB`
+            Pessimistic initialized pattern matcher object.
```
[the remark this answers is not in the document]
Comment (jdswinbank) [reply python/lsst/meas/astrom/matchPessimisticB.py]: Off the top of my head, I quite like `MatcherState`.

File: tests/test_matchPessimisticB.py [hunk 4]
```diff
@@ -147,6 +147,42 @@ def plot(catalog, symbol):
...
+
+        matchRes = self.MatchPessimisticB.matchObjectsToSources(
+            refCat=refCat,
+            sourceCat=sourceCat,
+            wcs=self.distortedWcs,
+            refFluxField="r_flux",
+            match_tolerance=matchTol,
+        )
+
+        self.assertEquals(len(matchRes.matches), 183)
```
[the remark this answers is not in the document]
Comment (morriscb) [reply tests/test_matchPessimisticB.py]: I think that 183 here is just the length of the input catalog as the unittests just distort and the catalog and match it to itself. The main test here is that all the variables in the output match_tolerance get set to some value and have the correct names. The last assert is a test that something hasn't gone terribly wrong. Do you have a suggestion of an assert that would catch what we want to test here? i.e. that match tolerance has been set properly. I could test that it returns an expected set of values in match_tolerance.

File: tests/test_matchPessimisticB.py [hunk 5]
```diff
@@ -147,6 +147,42 @@ def plot(catalog, symbol):
...
+        distortedCat = distort.distortList(sourceCat, distort.linearXDistort)
+
+        sourceCat = distortedCat
+
+        matchRes = self.MatchPessimisticB.matchObjectsToSources(
+            refCat=refCat,
+            sourceCat=sourceCat,
+            wcs=self.distortedWcs,
+            refFluxField="r_flux",
+        )
```
[the remark this answers is not in the document]
Comment (morriscb) [reply tests/test_matchPessimisticB.py]: The main test here is that all the variables in the output match_tolerance get set to some value and have the correct names. The last assert is a test that something hasn't gone terribly wrong. Do you have a suggestion of an assert that would catch what we want to test here? i.e. that match tolerance has been set properly. I could test that it returns an expected set of values in match_tolerance.

File: tests/test_matchPessimisticB.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply tests/test_matchPessimisticB.py]: I think 183 is just the length of the input catalog. If that's the case I could just replace all references to 183 with len(sourceCat).

[the remark this answers is not in the document]
Comment (jdswinbank) [reply tests/test_matchPessimisticB.py]: I like that suggestion.

File: tests/test_matchPessimisticB.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdswinbank) [reply tests/test_matchPessimisticB.py]: So I guess what you really want to show is that providing `match_tolerance` to `matchObjectsToSources` does _something_ — ie, that the second iteration of matching is different from the first. A couple of thoughts on that:

Are _any_ of the results of `matchRes` different on the second iteration? If you can show that `autoMaxMatchDist` or something has changed, then it shows that providing `match_tolerance` has done something, and it's mission accomplished.

Failing that, how about creating a `MatchTolerancePessimistic` object with known bad (but not catastrophic) values of these parameters (double the `autoMaxMatchDist`, or something), then demonstrating you get a “fixed” version back after you run `matchObjectsToSources`?

[the remark this answers is not in the document]
Comment (morriscb) [reply tests/test_matchPessimisticB.py]: Added a few tests. Take a look and see what you think.

File: tests/test_matchPessimisticB.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply tests/test_matchPessimisticB.py]: The length of the catalog is actually 183 so I now set the value in setup.

**Pull request merged**